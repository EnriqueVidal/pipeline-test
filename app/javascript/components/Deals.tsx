import * as React from 'react';
import Bar from './Bar';
import DealRow from './DealRow';
import useTabular, { Deal } from '../packs/useTabular';
import usePagination, { Pagination } from '../packs/usePagination';
import { navigate } from '../packs/navigation';

interface Props {
 deals: {
   entries: Deal[],
   pagination: Pagination,
 },
}

const BASE_API_URL = '/api/v1/deals';

const Deals = ({ deals }: Props) => {
  const {
    currentDeals, direction, setDeals, sortColumn,
  } = useTabular(deals.entries);

  const {
    changeLimit,
    moveBack,
    moveForward,
    pagination,
  } = usePagination(deals.pagination);

  const { page, per_page: perPage } = pagination;

  const sortableTile = `Sort ${direction}`;

  React.useEffect(() => {
    const abortController = new AbortController();

    fetch([BASE_API_URL, page, perPage].join('/'), {
      signal: abortController.signal,
    })
      .then(async (response) => {
        const { entries } = await response.json();

        if (abortController.signal.aborted) {
          return;
        }

        setDeals(entries);
      });

    return () => abortController.abort();
  }, [page, perPage]);

  const handleMoveBack = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();

    navigate(page - 1);
    moveBack();
  };

  const handleMoveForward = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();

    navigate(page + 1);
    moveForward();
  };

  const handleSelectLimit = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const { value } = event.target;
    changeLimit(parseInt(value, 10));
  };

  const handleSort = (columnName: string) => (event: React.MouseEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    sortColumn(columnName);
  };

  return (
    <div className="deals">
      <Bar
        handleMoveBack={handleMoveBack}
        handleMoveForward={handleMoveForward}
        handleSelectLimit={handleSelectLimit}
        pagination={pagination}
      />
      <div className="content">
        <table className="table is-striped is-fullwidth">
          <thead>
            <tr>
              <th>
                <a href="#" onClick={handleSort('id')} title={sortableTile}>ID</a>
              </th>
              <th>User</th>
              <th>Name</th>
              <th>
                <a href="#" onClick={handleSort('value')} title={sortableTile}>Value</a>
              </th>
              <th>Stage</th>
              <th>
                <a
                  href="#"
                  onClick={handleSort('deal_stage.percent')}
                  title={sortableTile}
                >
                  Progress
                </a>
              </th>
            </tr>
          </thead>
          <tbody>
            {currentDeals.map((deal) => <DealRow key={deal.id} deal={deal} />)}
          </tbody>
        </table>
      </div>
    </div>
  );
};

Deals.defaultProps = {
  deals: [],
};

export default Deals;
