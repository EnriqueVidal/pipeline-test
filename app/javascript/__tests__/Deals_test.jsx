import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Deals from '../components/Deals';

describe('Deals', () => {
  const spaceJam = {
    id: 1,
    name: 'SpaceJam',
    deal_stage: { id: 1, percent: 100, name: 'Won!' },
    user: { id: 1, first_name: 'Michael', last_name: 'Jordan' },
    value: '1000000000',
  };

  const pagination = {
    page: 1,
    per_page: 1,
    total: 2,
    pages: 1,
  };

  const firstPage = { entries: [spaceJam], pagination };

  global.fetch = jest.fn(async () => ({
    json: () => Promise.resolve(firstPage),
  }));

  it('matches snapshot', () => {
    const tree = renderer
      .create(<Deals deals={firstPage} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('mounts', () => {
    const wrapper = mount(<Deals deals={firstPage} />);
    expect(wrapper).not.toBeNull();
  });
});
