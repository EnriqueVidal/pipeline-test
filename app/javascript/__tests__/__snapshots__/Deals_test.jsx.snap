// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deals matches snapshot 1`] = `
<div
  className="deals"
>
  <div
    className="floating-bar columns is-mobile"
  >
    <div
      className="column"
    >
      <button
        className="button is-link is-small"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        &lt;
         
        Back
      </button>
    </div>
    <div
      className="column"
    >
      <div
        className="field"
      >
        <div
          className="control is-expanded"
        >
          <div
            className="select is-fullwidth is-small"
          >
            <select
              id="perPage"
              onChange={[Function]}
              value={1}
            >
              <option
                value={5}
              >
                5
              </option>
              <option
                value={10}
              >
                10
              </option>
              <option
                value={15}
              >
                15
              </option>
              <option
                value={30}
              >
                30
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
    <div
      className="column"
    >
      <div
        className="buttons is-right"
      >
        <button
          className="button is-link is-small"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Forward
           
          &gt;
        </button>
      </div>
    </div>
  </div>
  <div
    className="content"
  >
    <table
      className="table is-striped is-fullwidth"
    >
      <thead>
        <tr>
          <th>
            <a
              href="#"
              onClick={[Function]}
              title="Sort ASC"
            >
              ID
            </a>
          </th>
          <th>
            User
          </th>
          <th>
            Name
          </th>
          <th>
            <a
              href="#"
              onClick={[Function]}
              title="Sort ASC"
            >
              Value
            </a>
          </th>
          <th>
            Stage
          </th>
          <th>
            <a
              href="#"
              onClick={[Function]}
              title="Sort ASC"
            >
              Progress
            </a>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <strong>
              1
            </strong>
          </td>
          <td>
            Michael Jordan
          </td>
          <td>
            SpaceJam
          </td>
          <td
            className="has-text-right"
          >
            $
            1000000000.00
          </td>
          <td>
            <span
              className="tag is-primary is-info"
            >
              Won!
            </span>
          </td>
          <td>
            <progress
              className="progress is-info"
              max={100}
              title="100%"
              value={100}
            >
              100%
            </progress>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
